### PR TITLE
[Feature] Pagination component

### PR DESCRIPTION
**Pull request by Riippi**
## Related Issue

Closes [#237](https://jira.dvv.fi/browse/SFIDS-237)

## Motivation and Context
We want to have Pagination component available in the library

### Pagination
* Introduce `<Pagination>` component

### CODE REVIEW

File: src/core/Pagination/Pagination.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import { HtmlSpan, HtmlNav, HtmlDiv } from '../../reset';
+import { asPropType } from '../../utils/typescript';
+import { PageInput } from './PageInput/PageInput';
+import { Button } from '../Button/Button';
+
+const externalClassName = 'fi-link--external';
+
+export interface PaginationInputProps {
+  /** Page input action button label for screen readers  */
+  pageInputButtonText: string;
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:15]: Just buttonText might work now that this is under the PageInputProps.

File: src/core/Pagination/Pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  lastPage: number;
+  /** Returns the selected page number */
+  onChange: (page: number) => void;
+
+  /** Ref is forwarded to root element. Alternative for React `ref` attribute. */
+  forwardedRef?: React.RefObject<HTMLElement>;
+  /** Next page button label for screen readers  */
+  ariaNextButtonLabel: string;
+  /** Previous page button label for screen readers */
+  ariaPreviousButtonLabel: string;
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:47]: These names would probably be clearer if they were 'previousButtonAriaLabel' etc.

File: src/core/Pagination/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  /** Is there any need to have as prop */
+  asProp?: asPropType;
+  /** aria-label for pagination  */
+  'aria-label': string;
+  /** Function for page number indicator text */
+  textFunction?: (currentPage: number, lastPage: number) => string;
+  /** Use small screen styling */
+  smallScreen?: boolean;
+  /** Number of the current page */
+  currentPage: number;
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:36]: Is there a need to allow other than controlled use of the component? E.g. only giving the component the lastPage value based on the amount of items and using the changing page number of the pagination component to load a new set of items.

File: src/core/Pagination/Pagination.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+      <HtmlNav
+        {...passProps}
+        className={classnames(baseClassName, className, externalClassName, {
+          [paginationClassNames.smallScreen]: !!smallScreen,
+        })}
+      >
+        <HtmlDiv className={paginationClassNames.styleWrapper}>
+          <HtmlDiv className={paginationClassNames.buttonsWrapper}>
+            <Button
+              className={paginationClassNames.stockButton}
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:115]: Where does the `stockButton` name come from? Would `arrowButton` or something of the sort be clearer?

File: src/core/Pagination/Pagination.tsx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  /** Ref is forwarded to root element. Alternative for React `ref` attribute. */
+  forwardedRef?: React.RefObject<HTMLElement>;
+  /** Next page button label for screen readers  */
+  ariaNextButtonLabel: string;
+  /** Previous page button label for screen readers */
+  ariaPreviousButtonLabel: string;
+
+  pageInput?: boolean;
+
+  paginationInputProps?: PaginationInputProps;
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:51]: if the name of the element is `pageInput`, would `pageInputProps` make sense here?

File: src/core/Pagination/Pagination.tsx [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+              className={paginationClassNames.stockButton}
+              variant="secondary"
+              onClick={this.onRightButtonClick}
+              disabled={currentPage >= lastPage}
+              aria-label={ariaNextButtonLabel}
+              icon="arrowRight"
+            />
+          </HtmlDiv>
+
+          {pageInput === true && paginationInputProps && (
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.tsx:141]: I think the props should probably be handled via conditional props so that the page input props are required when `pageInput===true`. Now it's a bit confusing that the input doesn't render even if one sets `pageInput` to true.

File: src/core/Pagination/Pagination.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        labelText: 'Page number'
+      }}
+      onChange={(page) => {
+        setCurrent(page);
+        console.log('on change: ', page);
+      }}
+      textFunction={(currentPage, lastPage) => {
+        return 'Page ' + currentPage + ' / ' + lastPage;
+      }}
+      aria-label="my component here"
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.md:38]: Why is this here? Remnants of routerlink copypaste?

File: src/core/Pagination/Pagination.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    />
+    <br />
+    <br />
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.md:41]: Do these serve a purpose?

File: src/core/Pagination/Pagination.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      ariaPreviousButtonLabel="Previous"
+      paginationInputProps={{
+        invalidValueErrorText: ' is not allowed',
+        inputPlaceholderText: 'Go to',
+        pageInputButtonText: 'Jump to page',
+        labelText: 'Page number'
+      }}
+      onChange={(page) => {
+        setCurrent(page);
+        console.log('on change: ', page);
```
Comment (LJKaski) [inline src/core/Pagination/Pagination.md:33]: Probably can be removed when no longer needed?

File: src/core/Pagination/PageInput/PageInput.tsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+  state: PageInputState = {
+    value: '',
+    status: this.props.status || 'default',
+    inputValue: undefined,
+  };
+
+  private inputRef = createRef<HTMLInputElement>();
+
+  private getStatusText = () =>
+    `"${this.state.inputValue}" ${this.props.statusText}`;
```
Comment (LJKaski) [inline src/core/Pagination/PageInput/PageInput.tsx:105]: Should the statusText be a function that the developer can pass to allow more flexibility?

Comment (LJKaski) [reply src/core/Pagination/PageInput/PageInput.tsx:105]: Or simply leave this logic out and let the developer get the value and add it to the statusText if they want.

File: src/core/Pagination/Pagination.tsx (hunk 1 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: There is now separate PageInputProps interface

File: src/core/Pagination/Pagination.tsx (hunk 2 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: These are now renamed

File: src/core/Pagination/Pagination.tsx (hunk 4 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: `stockButton` was a remnant from earlier experiment. Renamed now.

File: src/core/Pagination/Pagination.tsx (hunk 6 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: This is now conditional prop

File: src/core/Pagination/Pagination.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    >
+      <Heading variant="h3"> Page: {current}</Heading>
+    </Block>
+    <br />
+    <Pagination
+      currentPage={current}
+      lastPage={lastPage}
+      pageInput={true}
+      smallScreen={false}
+      ariaNextButtonLabel="Next"
```
[the remark this answers is not in the document]
Comment (Riippi) [reply src/core/Pagination/Pagination.md]: Renamed

File: src/core/Pagination/Pagination.md (hunk 8 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.md]: No. Removed now.

File: src/core/Pagination/Pagination.md (hunk 9 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.md]: Yes. Removed now.

File: src/core/Pagination/Pagination.md (hunk 7 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.md]: Removed

File: src/core/Pagination/PageInput/PageInput.tsx (hunk 10 above)
Comment (Riippi) [reply src/core/Pagination/PageInput/PageInput.tsx]: There is now a function prop to handle this

File: src/core/Pagination/Pagination.tsx (hunk 3 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: Now `currentPage` is stored in component state

File: src/core/Pagination/PageInput/PageInput.baseStyles.tsx [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+      border: 0;
+      min-height: 36px;
+      margin-top: 1px;
+      margin-bottom: 1px;
+      &::placeholder {
+        font-style: italic;
+        color: ${theme.colors.depthDark2};
+        opacity: 1;
+      }
+      &::-ms-clear {
```
Comment (jenkrisu) [inline src/core/Pagination/PageInput/PageInput.baseStyles.tsx:68]: I tested with Legacy Edge and seems ms-clear & ms-reveal are just for IE, so not necessary anymore.

File: src/core/Pagination/PageInput/PageInput.baseStyles.tsx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        display: none;
+        width: 0;
+        height: 0;
+      }
+      &::-ms-reveal {
+        display: none;
+        width: 0;
+        height: 0;
+      }
+      &::-webkit-search-decoration,
```
Comment (jenkrisu) [inline src/core/Pagination/PageInput/PageInput.baseStyles.tsx:78]: Seems -webkit-search-... styles are only for inputs with type="search", so not needed when PageInput has type="text"

File: src/core/Pagination/Pagination.tsx [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+      pageInputProps: PageInputProps;
+    };
+
+interface InternalPaginationProps {
+  /** Custom classname to extend or customize */
+  className?: string;
+  /** Is there any need to have as prop */
+  asProp?: asPropType;
+  /** aria-label for pagination  */
+  'aria-label': string;
```
Comment (jenkrisu) [inline src/core/Pagination/Pagination.tsx:42]: Is this aria-label prop used somewhere?

**Comment by jenkrisu** [comment]
I noticed when testing PageInput, I can write number once, but then after clicking the PageInput button, I can't write to the field again. Doesn't happen with with Enter. Can you reproduce? 🤔

File: src/core/Pagination/Pagination.tsx (hunk 14 above)
Comment (jenkrisu) [reply src/core/Pagination/Pagination.tsx:42]: Also the asProp is maybe left over unnecessarily?

File: src/core/Pagination/PageInput/PageInput.baseStyles.tsx (hunk 13 above)
Comment (Riippi) [reply src/core/Pagination/PageInput/PageInput.baseStyles.tsx:78]: Removed now.

File: src/core/Pagination/Pagination.tsx [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+  onChange: (page: number) => void;
+  /** Ref is forwarded to root element. Alternative for React `ref` attribute. */
+  forwardedRef?: React.RefObject<HTMLElement>;
+  /** Next page button label for screen readers  */
+  nextButtonAriaLabel: string;
+  /** Previous page button label for screen readers */
+  previousButtonAriaLabel: string;
+}
+
+// export type PaginationProps = inputProps & InternalPaginationProps;
```
Comment (jenkrisu) [inline src/core/Pagination/Pagination.tsx:61]: Some old line comment?

File: src/core/Pagination/Pagination.tsx (hunk 14 above)
Comment (jenkrisu) [reply src/core/Pagination/Pagination.tsx:42]: asProp seems to still be there if I looked correctly so I unresolved this

File: src/core/Pagination/Pagination.tsx (hunk 15 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: Removed now.

File: src/core/Pagination/Pagination.tsx [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+  className?: string;
+  /** aria-label for pagination  */
+  'aria-label': string;
+  /**
+   * Unique id
+   * If no id is specified, one will be generated automatically
+   */
+  id?: string;
+  /** Function for page number indicator text */
+  pageIndicatorText?: (currentPage: number, lastPage: number) => string;
```
Comment (jenkrisu) [inline src/core/Pagination/Pagination.tsx:46]: Should pageIndicatorText be required since else it will be blank?

File: src/core/Pagination/Pagination.tsx [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+  /** Page indication text for screen readers */
+  ariaPageIndicatorText?: (currentPage: number, lastPage: number) => string;
```
Comment (jenkrisu) [inline src/core/Pagination/Pagination.tsx:48]: ariaPageIndicatorText should probably also be required, since otherwise screen reader won't read anything?

File: src/core/Pagination/Pagination.tsx (hunk 6 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: I think this was discussed at some meeting and this seemed like a common model in other components.

File: src/core/Pagination/Pagination.tsx (hunk 16 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: I suppose there is no reason to make it easy to use this without texts. Changed now.

File: src/core/Pagination/Pagination.tsx (hunk 17 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx]: Also changed now.

File: src/core/Pagination/PageInput/PageInput.tsx [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+  labelText: ReactNode;
+  /** Placeholder text for input. */
+  visualPlaceholder?: string;
+  /** Page button label for screen readers */
+  pageInputButtonLabel: string;
+  /** Callback for search button click */
+  onPageChange?: (value: PageInputValue) => void;
+  /** Maximum value */
+  maxValue: number;
+  /** Callback for search button click */
```
Comment (jenkrisu) [inline src/core/Pagination/PageInput/PageInput.tsx:46]: Should terminology here (onPageChange & invalidValueErrorText descriptions) be search button => page input button?

File: src/core/Pagination/PageInput/PageInput.tsx [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+                  ariaDescribedBy,
+                ])}
+                size={6}
+                aria-invalid={this.state.status === 'error'}
+                id={id}
+                forwardedRef={this.inputRef}
+                className={pageInputClassNames.inputElement}
+                value={this.state.value}
+                placeholder={visualPlaceholder}
+                onChange={(event: ChangeEvent<HTMLInputElement>) => {
```
Comment (jenkrisu) [inline src/core/Pagination/PageInput/PageInput.tsx:187]: Maybe onChange function block could be in its own function like onKeyPress and onKeyDown

File: src/core/Pagination/Pagination.tsx [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+type ShowInputProps =
+  | {
+      pageInput: false;
+      pageInputProps?: never;
+    }
+  | {
+      /** Show input field for page number */
+      pageInput?: true;
+      /** Props for page input field */
+      pageInputProps: PageInputProps;
```
Comment (jenkrisu) [inline src/core/Pagination/Pagination.tsx:32]: Can it be added to Styleguidist somehow, what this PageInputProps contains. For example <pre> tag with interface information in this comment?

File: src/index.tsx [hunk 21]
```diff
@@ -126,6 +126,7 @@ export {
   ExpanderTitleProps,
 } from './core/Expander/';
 export { Paragraph, ParagraphProps } from './core/Paragraph/Paragraph';
+export { Pagination, PaginationProps } from './core/Pagination/Pagination';
```
Comment (jenkrisu) [inline src/index.tsx:129]: Should PageInputProps also be imported?

File: src/core/Pagination/Pagination.tsx (hunk 20 above)
Comment (Riippi) [reply src/core/Pagination/Pagination.tsx:42]: This is now added with <pre> tags

File: src/index.tsx (hunk 21 above)
Comment (Riippi) [reply src/index.tsx]: Imported now

**Review by jenkrisu — APPROVED** [review]

**Review by jenkrisu — APPROVED** [review]
